### ISSUE TITLE
mypy ignores possible changes of global variables outside current function

### ISSUE THREAD

**Issue by lilydjwg** [python/mypy#11051]
**Bug Report**

See code.

**To Reproduce**

```py
from typing import Optional

class Global:
  attr: Optional[int] = None

G = Global()

def f() -> None:
  G.attr = 1

def t() -> None:
  G.attr = None
  f()
  assert G.attr is not None
  reveal_type(G)
```

1. Write the above code
2. Run mypy on it

**Expected Behavior**

Mypy shows the type of `G` at the last line (and `G.attr` is of type `int`).

**Actual Behavior**

Mypy completely skips the lines below `assert`.

```
Success: no issues found in 1 source file
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: no other args than the file
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.9.6
- Operating system and version: Arch Linux current

**Comment by KotlinIsland** [comment]
#11649
#11969

This issue has been discussed many times.
